neuraxle/pipeline.py

```diff
@@ -55,13 +60,73 @@ def inverse_transform(self, processed_outputs):
+    def __init__(self, cache_folder, pipeline_cache_list_file_name='pipeline_cache_list.txt'):
+        self.pipeline_cache_list_file_name = pipeline_cache_list_file_name
+        self.cache_folder = cache_folder
+
+    def save(self, pipeline: 'Pipeline', data_container: DataContainer):
+        pipeline_cache_folder = os.path.join(self.cache_folder, pipeline.name)
+
+        if not os.path.exists(pipeline_cache_folder):
+            os.makedirs(pipeline_cache_folder)
+
+        with open(os.path.join(pipeline_cache_folder, self.pipeline_cache_list_file_name), mode='a') as file:
+            next_cached_pipeline_path = self._create_next_cached_pipeline_path(pipeline_cache_folder)
+
+            file.write(str(data_container))
+            file.write(next_cached_pipeline_path)
+
+            dump(pipeline, next_cached_pipeline_path)
+
+        return pipeline
+
+    def load(self, pipeline: 'Pipeline', data_container: DataContainer):
+        pipeline_cache_folder = os.path.join(self.cache_folder, pipeline.name)
+        pipeline_cache_list_file_name_path = os.path.join(pipeline_cache_folder, self.pipeline_cache_list_file_name)
+        if not os.path.exists(pipeline_cache_list_file_name_path):
+            return pipeline
+
+        with open(pipeline_cache_list_file_name_path, mode='r') as file:
+            found_cached_pipeline = False
+            for line in file.readlines():
+                if found_cached_pipeline:
+                    return load(line)
+
+                if str(data_container) == line:
+                    found_cached_pipeline = True
+
+        return pipeline
+
+    def _create_next_cached_pipeline_path(self, pipeline_cache_folder):
+        cached_pipeline_paths = [path for path in glob.glob(os.path.join(pipeline_cache_folder, '*'))]
+        pipeline_file_name_index = 0
+
+        while self._create_cached_pipeline_path(pipeline_cache_folder,
+                                                pipeline_file_name_index) in cached_pipeline_paths:
+            pipeline_file_name_index += 1
+
+        return self._create_cached_pipeline_path(pipeline_cache_folder, pipeline_file_name_index)
+
+    def _create_cached_pipeline_path(self, pipeline_cache_folder, pipeline_file_name_index):
+        return os.path.join(pipeline_cache_folder,
+                            '{0}.joblib'.format(str(pipeline_file_name_index)))
+
+
 class Pipeline(BasePipeline, ResumableStepMixin):
     """
     Fits and transform steps after latest checkpoint
     """
 
-    def __init__(self, steps: NamedTupleList):
+    def __init__(self, steps: NamedTupleList, pipeline_repository: PipelineRepository):
         BasePipeline.__init__(self, steps=steps)
+        self.pipeline_repository = pipeline_repository
+
+    def save(self, data_container: DataContainer):
+        if self.parent_step is None:
+            self.pipeline_repository.save(self, data_container)
+        else:
+            self.parent_step.save(data_container)
```

Comment:
Will this trigger saving the next steps that weren't executed yet? that might cost a lot.

Reply:
This makes me get a new idea on when to initialize a step. We could save only the initialized steps! But make such that the steps be initialized only **just before** they first ever see some fit data. So a pipeline would only call `.initialize(...)` on its steps during fitting AND just before the step ever sees data. 

Do you like this idea? Would you have arguments against doing what I suggest?

Reply:
@guillaume-chevalier it would be more costly to save and load each steps individually. it is more performant to save the whole pipeline

Reply:
not sure we should overcomplicate the initialize. do we really need to only initialize when are about to go through them ? don't know maybe. if there is a costly operation in the setup, that would be great, but I really don't think that is supposed to happen.

Reply:
to be done in the next design iteration  idea

Reply:
to be discussed